### PR TITLE
Self Updater

### PR DESCRIPTION
**Pull request by harry1453**
This is ready to go other than the issue of code signing.

### CODE REVIEW

File: src/brs/Constants.java [hunk 1]
```diff
@@ -104,6 +104,8 @@
   public static final String WEBSITE = "website";
   public static final String PROTOCOL = "protocol";
 
+  public static final String UPDATE_DIR = "update/new/";
```
Comment (blankey1337) [inline src/brs/Constants.java]: Should this path be prefixed with `./` for good hygiene?

File: src/brs/selfupdater/SelfUpdater.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                ze = zipInputStream.getNextEntry();
+            }
+        }
+
+        // Copy the update script from our resources
+        String scriptFileName = "update" + (isWindows() ? ".bat" : ".sh");
+        File script = new File(scriptFileName);
+        script.createNewFile();
+        script.setExecutable(true);
+        try (InputStream sourceFile = getClass().getResourceAsStream("/selfupdater/"+ scriptFileName);
```
Comment (blankey1337) [inline src/brs/selfupdater/SelfUpdater.java]: Should this be path prefixed with `./`?

File: src/brs/selfupdater/SelfUpdater.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            } else {
+                throw new IOException("SHA-256 hashes not equal. Calculated " + Convert.toHexString(calculatedHash) + " and expected " + Convert.toHexString(release.getSha256()));
+            }
+        }
+
+        // Unzip
+        logger.info("Unzipping update...");
+        try (ZipInputStream zipInputStream = new ZipInputStream(new FileInputStream(new File(Constants.UPDATE_DIR, "package.zip")))) {
+            byte[] buffer = new byte[1024];
+            ZipEntry ze = zipInputStream.getNextEntry();
```
[the remark this answers is not in the document]
Comment (harry1453) [reply src/brs/selfupdater/SelfUpdater.java]: good thinking!

File: src/brs/selfupdater/SelfUpdater.java (hunk 2 above)
Comment (harry1453) [reply src/brs/selfupdater/SelfUpdater.java]: nope, it's accessing jar resources

File: src/brs/selfupdater/SelfUpdater.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        if (!Burst.getPropertyService().getBoolean(Props.BRS_SELF_UPDATER_ENABLE)) {
+            logger.info("Not starting auto-updater: Disabled in config.");
+            return;
+        }
+
+        if (Burst.getPropertyService().getBoolean(Props.BRS_SELF_UPDATER_ANYTIME)) {
+            logger.info("Starting self update thread!");
+            new Thread(() -> {
+                //noinspection InfiniteLoopStatement
+                while(true) checkForUpdate(false);
```
[the remark this answers is not in the document]
Comment (harry1453) [reply src/brs/selfupdater/SelfUpdater.java]: it's not very clear but it passes false to the function so one of the functions first few lines is a 6-24 hour delay. I admit it's not very readable but it's okay to have the loop in there

**Comment by harry1453** [comment]
Multisig is in there now

**Comment by harry1453** [comment]
This is code-complete, please review now

**Comment by harry1453** [comment]
Merging blocked awaiting CIP -> Community Approval.

**Comment by harry1453** [comment]
Oops. Looks like that happened automatically when I re-created develop branch.

**Comment by harry1453** [comment]
Closing due to conflicts now existing and awaiting CIP

**Pull request closed without merging**